Mathlib/NumberTheory/SmoothNumbers.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.Data.Nat.Factorization.Basic
+
+/-!
+# Smooth numbers
+
+We define the set `Nat.smoothNumbers n` consisting of the positive natural numbers all of
+whose prime factors are strictly less than `n`.
+
+We also define the finite set `Nat.primesBelow n` to be the set of prime numbers less than `n`.
+
+The main definition `Nat.equivProdNatSmoothNumbers` establishes the bijection between
+`ℕ × (smoothNumbers p)` and `smoothNumbers (p+1)` given by sending `(e, n)` to `p^e * n`.
+Here `p` is a prime number.
+-/
+
+namespace Nat
+
+/-- `primesBelow n` is the set of primes less than `n` as a finset. -/
+def primesBelow (n : ℕ) : Finset ℕ := (Finset.range n).filter (fun p ↦ p.Prime)
+
+@[simp]
+lemma primesBelow_zero : primesBelow 0 = ∅ := by
+  rw [primesBelow, Finset.range_zero, Finset.filter_empty]
+
+lemma prime_of_mem_primesBelow {p n : ℕ} (h : p ∈ n.primesBelow) : p.Prime :=
+  (Finset.mem_filter.mp h).2
+
+lemma lt_of_mem_primesBelow {p n : ℕ} (h : p ∈ n.primesBelow) : p < n :=
+  Finset.mem_range.mp <| Finset.mem_of_mem_filter p h
+
+lemma primesBelow_succ (n : ℕ) :
+    primesBelow n.succ = if n.Prime then insert n (primesBelow n) else primesBelow n := by
+  rw [primesBelow, primesBelow, Finset.range_succ, Finset.filter_insert]
+
+lemma not_mem_primesBelow (n : ℕ) : n ∉ primesBelow n :=
+  fun hn ↦ (lt_of_mem_primesBelow hn).false
+
+/-- `smoothNumbers n` is the set of *`n`-smooth positive natural numbers*, i.e., the
+positive natural numbers all of whose prime factors are less than `n`. -/
+def smoothNumbers (n : ℕ) : Set ℕ := {m | m ≠ 0 ∧ ∀ p ∈ factors m, p < n}
+
+lemma mem_smoothNumbers {n m : ℕ} : m ∈ smoothNumbers n ↔ m ≠ 0 ∧ ∀ p ∈ factors m, p < n :=
```

Comment:
maybe we want a lemma that says it's equivalent to `∀p, p | m => p < n`, that way zero can get excluded automatically too

Reply:
That statement is wrong: 4 is in `smoothNumbers 3` and `4 ∣ 4`, but `4 < 3` is false.

Reply:
for primes `p`?

Reply:
OK; that works. I'm adding
```lean
/-- `m` is `n`-smooth if and only if all prime divisors of `m` are less than `n`. -/
lemma mem_smoothNumbers' {n m : ℕ} : m ∈ smoothNumbers n ↔ ∀ p, p.Prime → p ∣ m → p < n := ...
```